src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryTable.cs

```diff
@@ -408,6 +408,76 @@ private static async Task<(DValue<RecordValue> row, FormulaValue sortValue)> App
             return (row, sortValue);
         }
 
+        private static async Task<(DValue<RecordValue> row, FormulaValue sortValue)> ApplyDistinctLambda(EvalVisitor runner, EvalVisitorContext context, DValue<RecordValue> row, LambdaFormulaValue lambda)
+        {
+            if (!row.IsValue)
+            {
+                return (row, row.ToFormulaValue());
+            }
+
+            var childContext = context.SymbolContext.WithScopeValues(row.Value);
+            var distinctValue = await lambda.EvalInRowScopeAsync(context.NewScope(childContext));
+
+            return (row, distinctValue);
+        }
+
+        public static async ValueTask<FormulaValue> DistinctTable(EvalVisitor runner, EvalVisitorContext context, IRContext irContext, FormulaValue[] args)
+        {
+            var arg0 = (TableValue)args[0];
+            var arg1 = (LambdaFormulaValue)args[1];
+
+            var pairs = (await Task.WhenAll(arg0.Rows.Select(row => ApplyDistinctLambda(runner, context, row, arg1)))).ToList();
```

Comment:
Could we have a well structured variable tree? Easier to debug via breakpoints.

Ex (not tested):

```cs
var values = arg0.Rows.Select(row => ApplyDistinctLambda(runner, context, row, arg1));
var pairResults = await Task.WhenAll(values);
var pairs = pairResults.ToList();
```